### PR TITLE
[circle-mpqsolver] Implement Dumper

### PR DESCRIPTION
**Pull request by stamalakhov**
This commit implements dumper for saving intermediate results.

Its correctness is tested in https://github.com/Samsung/ONE/pull/10655

Draft: https://github.com/Samsung/ONE/pull/10655
Related: https://github.com/Samsung/ONE/issues/10634

ONE-DCO-1.0-Signed-off-by: s.malakhov <s.malakhov@partner.samsung.com>

### CODE REVIEW

File: compiler/circle-mpqsolver/src/core/Dumper.cpp [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+namespace
+{
+const std::string default_dtype_key = "default_quantization_dtype";
+const std::string default_granularity_key = "default_granularity";
+const std::string layers_key = "layers";
+const std::string model_key = "model_path";
+const std::string layer_name_key = "name";
+const std::string layer_dtype_key = "dtype";
+const std::string layer_granularity_key = "granularity";
+} // namespace
```
Comment (seanshpark) [inline compiler/circle-mpqsolver/src/core/Dumper.cpp:37]:
```suggestion
namespace
{

const std::string default_dtype_key = "default_quantization_dtype";
const std::string default_granularity_key = "default_granularity";
const std::string layers_key = "layers";
const std::string model_key = "model_path";
const std::string layer_name_key = "name";
const std::string layer_dtype_key = "dtype";
const std::string layer_granularity_key = "granularity";

} // namespace
```

File: compiler/circle-mpqsolver/src/core/Dumper.cpp [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+  {
+    if (mkdir(dir_path.c_str(), S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH) != 0)
+    {
+      throw std::runtime_error("Failed to create directory for dumping intermediate results");
+    }
+  }
+}
+
+void Dumper::dump_MPQ_configuration(const LayerParams &layers, const std::string &def_dtype,
+                                    int param) const
```
Comment (jinevening) [inline compiler/circle-mpqsolver/src/core/Dumper.cpp:81]: Better to rename `param`. Maybe `step` or `iter_num`?
```suggestion
                                    int step) const
```

File: compiler/circle-mpqsolver/src/core/Dumper.cpp [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+{
+  luci::CircleExporter exporter;
+  luci::CircleFileExpContract contract(module, path);
+  if (!exporter.invoke(&contract))
+  {
+    throw std::runtime_error("Failed to export circle model to " + path);
+  }
+}
+
+void Dumper::dump_quantized(luci::Module *module, uint32_t param) const
```
Comment (jinevening) [inline compiler/circle-mpqsolver/src/core/Dumper.cpp:113]: Ditto.

```suggestion
void Dumper::dump_quantized(luci::Module *module, uint32_t step) const
```

**Review by jinevening — APPROVED** [review]
LGTM

**Comment by seanshpark** [comment]
@jinevening , please do not merge when there exist unconfirmed from the reviewer list.

**Comment by jinevening** [comment]
@seanshpark Sorry. It was my mistake. I merged in the app, and I was not familiar with app interface. I missed that there was another unconfirmed review. I'll double check the app interface. Sorry again.

**Pull request merged**